tiledb/sm/filesystem/gcs.cc

```diff
@@ -53,6 +60,551 @@ GCS::~GCS() {
+}
+
+Status GCS::remove_bucket(const URI& uri) const {
+  assert(client_);
+
+  if (!uri.is_gcs()) {
+    return LOG_STATUS(Status::GCSError(
+        std::string("URI is not a GCS URI: " + uri.to_string())));
+  }
+
+  // Empty bucket
+  RETURN_NOT_OK(empty_bucket(uri));
+
+  std::string bucket_name;
+  RETURN_NOT_OK(parse_gcs_uri(uri, &bucket_name, nullptr));
+
+  const google::cloud::Status status = client_->DeleteBucket(bucket_name);
+  if (!status.ok()) {
+    return LOG_STATUS(Status::GCSError(std::string(
+        "Delete bucket failed on: " + uri.to_string() + " (" +
+        status.message() + ")")));
+  }
+
+  return wait_for_bucket_to_be_deleted(bucket_name);
+}
+
+Status GCS::remove_object(const URI& uri) const {
+  assert(client_);
+
+  if (!uri.is_gcs()) {
+    return LOG_STATUS(Status::GCSError(
+        std::string("URI is not a GCS URI: " + uri.to_string())));
+  }
+
+  std::string bucket_name;
+  std::string object_path;
+  RETURN_NOT_OK(parse_gcs_uri(uri, &bucket_name, &object_path));
+
+  const google::cloud::Status status =
+      client_->DeleteObject(bucket_name, object_path);
+  if (!status.ok()) {
+    return LOG_STATUS(Status::GCSError(std::string(
+        "Delete object failed on: " + uri.to_string() + " (" +
+        status.message() + ")")));
+  }
+
+  return wait_for_object_to_be_deleted(bucket_name, object_path);
+}
+
+Status GCS::remove_dir(const URI& uri) const {
+  assert(client_);
+
+  if (!uri.is_gcs()) {
+    return LOG_STATUS(Status::GCSError(
+        std::string("URI is not a GCS URI: " + uri.to_string())));
+  }
+
+  std::vector<std::string> paths;
+  RETURN_NOT_OK(ls(uri, &paths, ""));
+  for (const auto& path : paths) {
+    RETURN_NOT_OK(remove_object(URI(path)));
+  }
+
+  return Status::Ok();
+}
+
+std::string GCS::remove_front_slash(const std::string& path) const {
+  if (path.front() == '/') {
+    return path.substr(1, path.length());
+  }
+
+  return path;
+}
+
+std::string GCS::add_trailing_slash(const std::string& path) const {
+  if (path.back() != '/') {
+    return path + "/";
+  }
+
+  return path;
+}
+
+std::string GCS::remove_trailing_slash(const std::string& path) const {
+  if (path.back() == '/') {
+    return path.substr(0, path.length() - 1);
+  }
+
+  return path;
+}
+
+Status GCS::ls(
+    const URI& uri,
+    std::vector<std::string>* paths,
+    const std::string& delimiter,
+    const int max_paths) const {
+  assert(client_);
+  assert(paths);
+
+  const URI uri_dir = uri.add_trailing_slash();
+
+  if (!uri_dir.is_gcs()) {
+    return LOG_STATUS(Status::GCSError(
+        std::string("URI is not a GCS URI: " + uri_dir.to_string())));
+  }
+
+  std::string bucket_name;
+  std::string object_path;
+  RETURN_NOT_OK(parse_gcs_uri(uri_dir, &bucket_name, &object_path));
+
+  google::cloud::storage::Prefix prefix_option(object_path);
+
+  google::cloud::storage::ListObjectsReader objects_reader =
+      client_->ListObjects(bucket_name, std::move(prefix_option));
```

Comment:
Is there any way to set a maximum number of results returned?

Reply:
Not immediately in the SDK. We could consider patching it if the HTTP request allows it. This is now less concerning because the last patch added support for a delimiter -- so we won't be fetching extra objects.